### PR TITLE
fixes for #140 and #165

### PR DESCRIPTION
**Pull request by amarzavery**
fixes for #140 and #165

### CODE REVIEW

**Comment by amarzavery** [comment]
@vishrutshah and @veronicagg please review.

File: test/liveValidatorTests.js [hunk 1]
```diff
@@ -117,7 +117,7 @@ describe('Live Validator', function () {
       let validator = new LiveValidator(options);
       validator.initialize().then(function () {
         assert.equal(true, expectedProvider in validator.cache);
-        assert.equal(5, Object.keys(validator.cache).length);
+        assert.equal(6, Object.keys(validator.cache).length);
```
Comment (vishrutshah) [inline test/liveValidatorTests.js]: QQ: Is this because of the test.json or the fix ?

Comment (amarzavery) [reply test/liveValidatorTests.js]: yes. I could have added this in a separate folder. But then thought of keeping all the swaggers together. In that way we know that live validator will work well for the newly added edge case scenario tests as well.

File: lib/validators/specValidator.js [hunk 2]
```diff
@@ -603,22 +596,35 @@ class SpecValidator {
...
       let location = parameter.in;
       if (location === 'path' || location === 'query') {
+        if (location === 'path' && parameterValue && typeof parameterValue.valueOf() === "string") {
+          if (options.pathTemplate.charAt(options.pathTemplate.indexOf(`${parameter.name}`) - 2) === '/' && parameterValue.startsWith('/')) {
+            let msg = `In operation "${operation.operationId}", example for parameter "${parameter.name}": "${parameterValue}" starts with a forward slash ` +
+              `and the path template: "${options.pathTemplate}" contains a forward slash before the parameter starts. This will cause double forward slashes ` +
+              ` in the request url. Thus making it incorrect. Please rectify the example.`;
+            let e = self.constructErrorObject(ErrorCodes.DoubleForwardSlashesInUrl, msg);
+            result.validationResult.errors.push(e);
+            foundIssues = true;
```
Comment (amarzavery) [inline lib/validators/specValidator.js]: I am still iffy about this. Not sure if throwing an error for this scenario is necessary.

Comment (vishrutshah) [reply lib/validators/specValidator.js]: I think throwing would give more visibility than swallowing...

File: lib/validators/specValidator.js [hunk 3]
```diff
@@ -603,22 +596,35 @@ class SpecValidator {
       }
       let location = parameter.in;
       if (location === 'path' || location === 'query') {
+        if (location === 'path' && parameterValue && typeof parameterValue.valueOf() === "string") {
+          if (options.pathTemplate.charAt(options.pathTemplate.indexOf(`${parameter.name}`) - 2) === '/' && parameterValue.startsWith('/')) {
```
Comment (veronicagg) [inline lib/validators/specValidator.js]: what's "-2" looking for here? it'd be great to add a small comment potentially with an example on why we're adding this condition. There's a message you're constructing right after, but a comment wouldn't hurt for readability.

File: lib/validators/specValidator.js [hunk 4]
```diff
@@ -603,22 +596,35 @@ class SpecValidator {
...
+            let msg = `In operation "${operation.operationId}", example for parameter "${parameter.name}": "${parameterValue}" starts with a forward slash ` +
+              `and the path template: "${options.pathTemplate}" contains a forward slash before the parameter starts. This will cause double forward slashes ` +
+              ` in the request url. Thus making it incorrect. Please rectify the example.`;
+            let e = self.constructErrorObject(ErrorCodes.DoubleForwardSlashesInUrl, msg);
+            result.validationResult.errors.push(e);
+            foundIssues = true;
+            break;
+          }
+          //replacing forward slashes with empty string because this messes up Sways regex validation of path segment.
+          parameterValue = parameterValue.replace(/\//ig, '');
```
Comment (veronicagg) [inline lib/validators/specValidator.js]: I found https://github.com/OAI/OpenAPI-Specification/issues/892. It appears that having path parameters with slashes may not be allowed in swagger??

File: lib/validators/specValidator.js (hunk 2 above)
Comment (veronicagg) [reply lib/validators/specValidator.js]: If we throw we stop any other validation, right? If we log an errors, we report it as any other? If so, I'd go with error.

File: test/swaggers/specification/scenarios/resource-manager/Microsoft.Test/2016-01-01/examples/storageAccountPathParameterForwardSlash.json [hunk 5]
```diff
@@ -0,0 +1,12 @@
+{
+  "parameters":{
+    "api-version": "2016-01-01",
+    "subscriptionId": "foo/foo1/foo2",
```
Comment (veronicagg) [inline test/swaggers/specification/scenarios/resource-manager/Microsoft.Test/2016-01-01/examples/storageAccountPathParameterForwardSlash.json]: we could make it a little bit more real, if we had another parameter other than subscriptionId, since subscriptionID does not follow this pattern.

File: test/swaggers/specification/scenarios/resource-manager/Microsoft.Test/2016-01-01/test.json [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+  "paths": {
+    "/subscriptions/{subscriptionId}/providers/Microsoft.Test/checkNameAvailability": {
+      "post": {
+        "tags": [
+          "StorageAccounts"
+        ],
+        "operationId": "StorageAccounts_CheckNameAvailability",
+        "description": "Checks that the storage account name is valid and is not already in use.",
+        "x-ms-examples": {
+          "storageAccountCheckNameAvailability": {"$ref": "./examples/storageAccountCheckNameAvailability.json"}
```
Comment (veronicagg) [inline test/swaggers/specification/scenarios/resource-manager/Microsoft.Test/2016-01-01/test.json]: you could probably have multiple examples for 2 operations (one for the forward slashes, one of the "?"), instead of multiple operations with one example each, it may simplify test.json.

File: lib/validators/specValidator.js (hunk 2 above)
Comment (amarzavery) [reply lib/validators/specValidator.js]: I think my sentence in the comment confused you guys. Sorry about that
> I am still iffy about this. Not sure if throwing an error for this scenario is necessary.

Yes we are just logging the error like we do for other scenarios. We are not throwing and halting further validation. 
What I actually meant to ask is 
> "Does it actually make sense to catch these kind of issues? I mean in the real world customers could provide forward slashes in their parameter values when they use an sdk or the portal or REST API directly. Shouldn't this be handled by the service. (they should ignore duplicate forward slashes or put a regex validation constraint in swagger spec that the value should not start with forward slash". Making this as a brute force validation in the model validator is too strict.

What do you think?

File: lib/validators/specValidator.js (hunk 4 above)
Comment (amarzavery) [reply lib/validators/specValidator.js]: Yes they are technically not allowed. But in real world services have such urls and we need to support them.
Think about apis like dropbox or file storage apis where they have a file path or container path as a path parameter. They wont be able to use swagger.

File: test/swaggers/specification/scenarios/resource-manager/Microsoft.Test/2016-01-01/test.json (hunk 6 above)
Comment (amarzavery) [reply test/swaggers/specification/scenarios/resource-manager/Microsoft.Test/2016-01-01/test.json]: I wanted to provide a scenario that has both the issues and verify whether our tool correctly catches it.

**Comment by amarzavery** [comment]
https://github.com/amarzavery/sway/pull/9

File: lib/validators/specValidator.js (hunk 2 above)
Comment (veronicagg) [reply lib/validators/specValidator.js]: yeah, I thought about that while reviewing too, I think we could just swallow honestly for now and fix it "silently", I think having a warning instead of an error could be a good thing, in case we decide to use this value for something else later and the impact is different - we would at least know it's flagged.
